### PR TITLE
Python Security Updates for 3.7.12

### PR DESCRIPTION
**Pull request by spham-amzn**
Updates to address CVEs for the latest 3.7.12 patch

* Expat update
* BZIP2 Update
* PIP Update

Also address issue of libffi being dependent on the system libffi.

https://github.com/o3de/o3de/issues/7764

### CODE REVIEW

**Review by lawsonamzn — APPROVED** [review]
LGTM. Some of these we offer already as downloadable packages though, maybe at some point we can use the packages.

File: package-system/python/linux_x64/make-python.sh [hunk 1]
```diff
@@ -46,18 +46,100 @@ echo ""
 mkdir -p temp
 
 
-echo ------------------------ GIT CLONE python 3.7 --------------------
+echo ""
+echo "--------------- GIT CLONE python 3.7 ---------------"
```
Comment (lmbr-pip) [inline package-system/python/linux_x64/make-python.sh:50]: nit: For other platforms you are echoing: "Cloning python 3.7.12 from git"

File: package-system/python/darwin_x64/make-python.sh [hunk 2]
```diff
@@ -34,34 +37,52 @@ SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
...
 retVal=$?
 if [ $retVal -ne 0 ]; then
-    echo "Error cloning python!"
+    echo "Error cloning python from https://github.com/python/cpython.git"
     exit $retVal
 fi
 
-echo "-------------- Cloning relocatable-python from git --------------"
+echo ""
+echo "---------------- Cloning expat 2.4.6 from git and applying update ----------------"
```
Comment (lmbr-pip) [inline package-system/python/darwin_x64/make-python.sh:72]: nit: Use libexpat to be consistent?

File: package-system/python/linux_x64/make-python.sh [hunk 3]
```diff
@@ -46,18 +46,100 @@ echo ""
...
 cd temp
 git clone https://github.com/python/cpython.git --branch v3.7.12 --depth 1
 
 if [[ ! -d "cpython" ]]; then
     echo "Was unable to create cpython dir via git clone.  Is git installed?"
     exit 1
 fi
+
+echo ""
+echo "--------------- Cloning expat 2.4.6 from git and applying update ---------------"
```
Comment (lmbr-pip) [inline package-system/python/linux_x64/make-python.sh:61]: nit: Use libexpat to be consistent?

File: package-system/python/win_x64/open3d_python.patch [hunk 4]
```diff
@@ -1,15 +1,108 @@
...
+index 4fab488..e8a0f5a 100755
+--- a/Mac/BuildScript/build-installer.py
++++ b/Mac/BuildScript/build-installer.py
+@@ -331,7 +331,7 @@ def library_recipes():
+     if getDeptargetTuple() < (10, 5):
+         result.extend([
+           dict(
+-              name="Bzip2 1.0.6",
++              name="Bzip2 1.0.8",
+               url="http://bzip.org/1.0.6/bzip2-1.0.6.tar.gz",
```
Comment (lmbr-pip) [inline package-system/python/win_x64/open3d_python.patch:29]: url appears to be 1.0.6 version, should be 1.0.8 version

**Review by lmbr-pip — APPROVED** [review]
LG.

I do find it odd we are hardcoding python, bzip versions everywhere (log statements, paths etc). Did we consider have constants with current versions so we just need to flip things once per .sh/.bat? Obviously not so easy for patch files where this would be hardcoded.

File: package-system/python/win_x64/open3d_python.patch (hunk 4 above)
Comment (spham-amzn) [reply package-system/python/win_x64/open3d_python.patch:29]: It turns out this file doesnt need patching at all. This is the patch for a windows build of python, but this section was patching the mac scripts, which will never be called anyways. Will remove this section

**Comment by Jackie9527** [comment]
For python-3.7.12-rev2-windows

Sqlite3 has CVEs as follows:
* [CVE-2021-36690](https://nvd.nist.gov/vuln/detail/CVE-2021-36690)

Tcl 8.6.11 has CVEs as follows:
* [CVE-2021-35331](https://nvd.nist.gov/vuln/detail/CVE-2021-35331)


python-3.7.12-rev2-linux is great, no more CVEs found currently.

**Comment by spham-amzn** [comment]
image

image

**Comment by Jackie9527** [comment]
LGTM
@spham-amzn

**Review by Jackie9527 — APPROVED** [review]
LGTM

**Comment by spham-amzn** [comment]
> > > For python-3.7.12-rev2-windows
> > > Sqlite3 has CVEs as follows:
> > > ```
> > > * [CVE-2021-36690](https://nvd.nist.gov/vuln/detail/CVE-2021-36690)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Tcl 8.6.11 has CVEs as follows:
> > > ```
> > > * [CVE-2021-35331](https://nvd.nist.gov/vuln/detail/CVE-2021-35331)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > python-3.7.12-rev2-linux is great, no more CVEs found currently.
> > 
> > 
> > How were those versions identified. This patch applies versions that are newer than the ones described in the CVEs
> > ```
> > sqlite-3.37.2.0
> > tcl-core-8.6.12.0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > Applying patch file D:\github\3p-package-source\package-system\python\win_x64\\open3d_python.patch
> > Getting external libraries
> > Using "D:\github\o3de\python\runtime\python-3.7.12-rev1-windows\python\python.exe" (from environment)
> > Fetching external libraries...
> > Fetching bzip2-1.0.8...
> > Fetching sqlite-3.37.2.0...
> > Fetching xz-5.2.2...
> > Fetching zlib-1.2.11...
> > Fetching external binaries...
> > Fetching openssl-bin-1.1.1m...
> > Fetching tcltk-8.6.12.0...
> > Finished.
> > ```
> 
> I think there might be something wrong. @spham-amzn

Wrong with which part?

**Review by lawsonamzn — APPROVED** [review]
LGTM. 

The packages of those softwares are newer than the CVEs.  Also, not all CVEs are legitimate and verified.  Both of those are disputed by the maintainers of the package with good reason.

**Review by Kadino — APPROVED** [review]

**Comment by Jackie9527** [comment]
> Wrong with which part? What would be helpful is transparency on how these CVEs are identified in the package

I mean it's my fault.

**Review by lmbr-pip — APPROVED** [review]

**Pull request merged**